drivers: usb: device: Fix MCUX usb performance issues

decsny opened the pull request:
This reverts commit https://github.com/zephyrproject-rtos/zephyr/commit/4e6f80d37a8de3b10da74f4356439c4d43c3fba2.

The reason this commit is being reverted is because it removed
an intermediate buffer used during the USB write operation.

This broke some platforms using flexspi when a USB descriptor
was stored in flash. When USB DMA as an AHB master tried to fetch
these descriptors from the flash through flexspi, there was a
performance hit causing the USB to miss a deadline in the protocol
and forced to return all 0's instead of the correct data.

The problem can be fixed by having an intermediate buffer in RAM
that the USB DMA can fetch from with minimal latency.

~~This PR also adds a Kconfig to control the size of this buffer for each endpoint,
since it was determined in the past that limiting the size of the transfer to
the ep->mps caused some issues, so it works better to have a fixed buffer size.~~

This PR also changes the way that the heap memory pool size is defined for this driver,
to ensure that there is enough space on the heap to allocate all the buffers for each endpoint.

tmon-nordic commented on drivers/usb/device/Kconfig:
```diff
@@ -142,6 +142,22 @@ config USB_MCUX
+config USB_MCUX_ENDPOINT_BUFFER_SIZE
+	int "Size of intermediate buffer in USB driver"
+	default 512
+	range 64 512
```
Is 512 the maximum MCUX can use? The USB 2.0 allows up to 1024 on high-speed interrupt endpoint.

decsny replied:
okay, I changed this

carlescufi commented:
@jfischer-no @tmon-nordic please review

jfischer-no commented:
I am confused by "This broke some platforms using flexspi when a USB descriptor was stored in flash.". Sure, there could be some data from flash, but I doubt it is USB descriptor.

jfischer-no commented on drivers/usb/device/Kconfig:
```diff
@@ -142,6 +142,22 @@ config USB_MCUX
 
+config USB_MCUX_ENDPOINT_BUFFER_SIZE
+	int "Size of intermediate buffer in USB driver"
+	default 512
```
I have not had time to look at it in detail, but this Kconfig option is problematic because it is not maximum by default, and a platform using this driver could require configuration overlay for some sample, subsequently I would block such overlay.

decsny replied:
I am curious what your rationale is for blocking a sample configuration overlay, but I will change this anyways

decsny replied:
actually I think I am just going to remove this Kconfig

jfischer-no commented on drivers/usb/device/Kconfig:
```diff
@@ -142,6 +142,22 @@ config USB_MCUX
+	  from missed deadlines when the descriptor is in some area that
+	  has high latency to the USB DMA. The size will be used for a
+	  buffer for each endpoint. It is not recommended to change this
+	  unless the maximum size of all USB descriptors is known in advance.
```
I really doubt it is descriptor issue, descriptors are placed in RAM section.

decsny replied:
I may be misunderstanding something but in this case it was `hid_report_desc` in the [HID sample](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/usb/hid/src/main.c), which was put into flash memory because it is declared const, which is why there was a flexspi performance issue and led to this PR.

jfischer-no commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -275,7 +275,7 @@ int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data *const cfg)
 	}
 
-	block->data = k_heap_alloc(&ep_buf_pool, cfg->ep_mps, K_NO_WAIT);
+	block->data = k_heap_alloc(&ep_buf_pool, CONFIG_USB_MCUX_ENDPOINT_BUFFER_SIZE, K_NO_WAIT);
```
Pool is defined by `K_HEAP_DEFINE(ep_buf_pool, 1024 * EP_BUF_NUMOF_BLOCKS);`
Can we remove K_HEAP_DEFINE and k_heap_alloc(), change `struct k_mem_block block;` to `uint8_t *block;` and define/assign static buffers to `s_ep_ctrl[]` using `LISTIFY(NUM_OF_EP_MAX, ....)`?

decsny replied:
I did have the same thought but I wasn't sure if it was appropriate for this PR, I suppose I can knock out two birds with one stone though?

decsny commented:
updated with @mmahadevan108's suggestion to exclude platforms using USB RAM from copying to an intermediate buffer

Commit:
drivers: usb_dc_mcux: Fix write performance issues

The MCUX USB device driver currently suffers from
some performance related issues caused by the removal
of the intermediate buffer in commit 4e6f80d37a8.
This buffer was essential in the case when some USB
descriptors or data were in a slow access memory,
for example a flash memory accessed over flexspi.
USB DMA as AHB master would try to fetch the data,
but it would be too slow and USB peripheral
could not meet deadlines on the USB bus waiting for it,
and so it would send null packets instead causing errors.

This problem can be fixed by re-introducing an intermediate
buffer in RAM with the data copied from the slow memory by
the core before the USB transfer begins, so that USB DMA
is not responsible for fetching it from flexspi/flash.

Changes to MCUX USB Device Driver:
- Re-add intermediate buffer for USB writes in RAM
- Buffer is not needed for reads, add runtime check
  to avoid copy overhead
- Buffer is not needed for platforms with USB RAM,
  since the USB RAM itself acts as an intermediate buffer.
  Compile with buffer code only when USB RAM is not present on
  the platform, to avoid unecessary copy overhead.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

mmahadevan108 commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -503,19 +515,38 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
+	buffer = (uint8_t *)dev_state.eps[ep_abs_idx].block.data;
+
+	if (data_len > CONFIG_USB_MCUX_ENDPOINT_BUFFER_SIZE) {
+		LOG_ERR("Packet too large");
```
The buffer is allocated for a size of `cfg->ep_mps`. However the check is done against `CONFIG_USB_MCUX_ENDPOINT_BUFFER_SIZE`. We probably don't need this Kconfig

decsny replied:
yes I had this on my local branch, I forgot to push

danieldegrasse commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -268,8 +274,10 @@ int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data *const cfg)
 	}
 
+#ifdef FSL_FEATURE_USBHSD_USB_RAM_BASE_ADDRESS
 	/* Allocate buffers used during read operation */
```
Minor question- should this comment state "Allocate buffers used during read operation"? Based on my understanding of USB RAM, it seems it is only used by the USB write operation.

decsny replied:
out direction is relative to the host, as this is a driver for a device, "out" would be a read
